targets/TARGET_STM/TARGET_STM32F4/sdio_api.c

```diff
@@ -0,0 +1,439 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include "sdio_api.h"
+#include "platform/mbed_error.h"
+
+
+#if DEVICE_SDIO
+#define CMD_TIMEOUT 30000
+/* Extern variables ---------------------------------------------------------*/
+
+SD_HandleTypeDef hsd;
+DMA_HandleTypeDef hdma_sdio_rx;
+DMA_HandleTypeDef hdma_sdio_tx;
+
+// simple flags for DMA pending signaling
+volatile int SD_DMA_ReadPendingState = SD_TRANSFER_OK;
+volatile int SD_DMA_WritePendingState = SD_TRANSFER_OK;
+
+/* DMA Handlers are global, there is only one SDIO interface */
+
+/**
+* @brief This function handles SDIO global interrupt.
+*/
+void _SDIO_IRQHandler(void)
+{
+    HAL_SD_IRQHandler(&hsd);
+}
+
+/**
+* @brief This function handles DMAx stream_n global interrupt. DMA Rx
+*/
+void _DMA_Stream_Rx_IRQHandler(void)
+{
+    HAL_DMA_IRQHandler(hsd.hdmarx);
+}
+
+/**
+* @brief This function handles DMAx stream_n global interrupt. DMA Tx
+*/
+void _DMA_Stream_Tx_IRQHandler(void)
+{
+    HAL_DMA_IRQHandler(hsd.hdmatx);
+}
+
+/**
+ *
+ * @param hsd:  Handle for SD handle Structure definition
+ */
+void HAL_SD_MspInit(SD_HandleTypeDef *hsd)
+{
+    IRQn_Type IRQn;
+    GPIO_InitTypeDef GPIO_InitStruct;
+
+    if (hsd->Instance == SDIO)
+    {
+        /* Peripheral clock enable */
+        __HAL_RCC_SDIO_CLK_ENABLE();
+        __HAL_RCC_DMA2_CLK_ENABLE();
+
+        /* Enable GPIOs clock */
+        __HAL_RCC_GPIOC_CLK_ENABLE();
+        __HAL_RCC_GPIOD_CLK_ENABLE();
+
+        /**SDIO GPIO Configuration
+         PC12     ------> SDIO_CK
+         PC11     ------> SDIO_D3
+         PC10     ------> SDIO_D2
+         PD2     ------> SDIO_CMD
+         PC9     ------> SDIO_D1
+         PC8     ------> SDIO_D0
+         */
+        GPIO_InitStruct.Pin = GPIO_PIN_12 | GPIO_PIN_11 | GPIO_PIN_10 | GPIO_PIN_9 | GPIO_PIN_8;
```

Comment:
This pin configuration can't be defined for all STM32 like this.
We should use some MACRO which can be define at each MCU level?
Or we should create a new structure in PeripheralPins.c ?

But OK to keep it like this in a first step, it can be optimized later while adding more SDIO targets.